### PR TITLE
proxy: listen on additional addrs

### PR DESCRIPTION
**Pull request by xhebox**
### What problem does this PR solve?

Issue Number: close #389 

Problem Summary: As title, requested by serverless team.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Notable changes

- [ ] Has configuration change
- [ ] Has HTTP API interfaces change
- [ ] Has tiproxyctl change
- [ ] Other user behavior changes

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Support listening SQL server on multiple addrs
```

### CODE REVIEW

**Comment by xuanyu66** [comment]
@xhebox Could we also add label `port` in gauge metrics?

File: lib/config/proxy.go [hunk 1]
```diff
@@ -60,10 +60,12 @@ type ProxyServerOnline struct {
 }
 
 type ProxyServer struct {
-	Addr              string `yaml:"addr,omitempty" toml:"addr,omitempty" json:"addr,omitempty"`
-	PDAddrs           string `yaml:"pd-addrs,omitempty" toml:"pd-addrs,omitempty" json:"pd-addrs,omitempty"`
-	ServerVersion     string `yaml:"server-version,omitempty" toml:"server-version,omitempty" json:"server-version,omitempty"`
-	RequireBackendTLS bool   `yaml:"require-backend-tls,omitempty" toml:"require-backend-tls,omitempty" json:"require-backend-tls,omitempty"`
+	Addr string `yaml:"addr,omitempty" toml:"addr,omitempty" json:"addr,omitempty"`
+	// Addrs will attach server to additional addrs.
+	Addrs             []string `yaml:"addrs,omitempty" toml:"addrs,omitempty" json:"addrs,omitempty"`
```
Comment (djshow832) [inline lib/config/proxy.go:65]: How about reusing `Addr` and separating the addresses by `,`? It's compatible with existing configs.
If we do add `Addrs`, we need to describe it in the official docs. I don't want that...

File: pkg/proxy/proxy.go [hunk 2]
```diff
@@ -66,10 +68,8 @@ func NewSQLServer(logger *zap.Logger, cfg config.ProxyServer, certMgr *cert.Cert
 
 	s.reset(&cfg.ProxyServerOnline)
 
-	s.addrs = make([]string, 0, len(cfg.Addrs)+1)
-	s.addrs = append(s.addrs, cfg.Addr)
-	s.addrs = append(s.addrs, cfg.Addrs...)
-
+	s.addrs = strings.Split(cfg.Addr, ",")
+	fmt.Printf("xhe %s\n", s.addrs)
```
Comment (zhangyangyu) [inline pkg/proxy/proxy.go:72]: forget to remove?

Comment (xhebox) [reply pkg/proxy/proxy.go]: Yep.

File: pkg/metrics/grafana/tiproxy_summary.json [hunk 3]
```diff
@@ -263,7 +263,7 @@
                      "expr": "tiproxy_server_connections{k8s_cluster=\"$k8s_cluster\", tidb_cluster=\"$tidb_cluster\", instance=~\"$instance\"}",
                      "format": "time_series",
                      "intervalFactor": 2,
-                     "legendFormat": "{{instance}}",
+                     "legendFormat": "{{instance}} | {{addr}}",
```
Comment (djshow832) [inline pkg/metrics/grafana/tiproxy_summary.json:266]: I don't think we need to adapt to that: Gateway doesn't use TiProxy metrics.

**Review by djshow832 — COMMENTED** [review]
Rest LGTM

File: pkg/metrics/grafana/tiproxy_summary.json (hunk 3 above)
Comment (djshow832) [reply pkg/metrics/grafana/tiproxy_summary.json:266]: Oops, I just saw the comment https://github.com/pingcap/tiproxy/pull/393#issuecomment-1788751694

**Comment by djshow832** [comment]
> @xhebox Could we also add the label `port` in gauge metrics? https://github.com/pingcap/tiproxy/pull/393/files#diff-b41d02f71c670ff801d99892d32ef6dd36e43fcdf4c12c044cd34a073cc18200L165

Gateway doesn't use TiProxy metrics. Are you going to add the metrics in Gateway?

**Comment by xuanyu66** [comment]
@djshow832 We want to see the traffic for each port. Do you mean that the gateway won't expose Tipoxy's metrics?

**Comment by djshow832** [comment]
> @djshow832 We want to see the traffic for each port. Do you mean that the gateway won't expose Tipoxy's metrics? Even if the gateway doesn't use these metrics, this change is meaningful for other users.

Yes, I mean Gateway doesn't expose those metrics and we have no way to see them.
This feature is only for Gateway. I don't think it's meaningful in other situations.

**Comment by xuanyu66** [comment]
@djshow832 LGTM

**Review by djshow832 — APPROVED** [review]

**Review by zhangyangyu — APPROVED** [review]

**Pull request merged**